### PR TITLE
fix rchain-3348. add measure and journaling to nextgen rspace

### PR DESCRIPTION
**Pull request by kayvank**
## Overview
migrate measure journal functionality form old rspace to nextgen rspace

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3348

### Notes
Created a new class, History Transformers, to create the required case class for the measure journal and moved the common transformers from HistoryRepositoryImpl to this new class

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    hashWithSuffix(channelsBits, continuationSuffixBits)
+  }
+
+  def hashDataChannel(channel: C): Blake2b256Hash =
+    hashWithSuffix(codecC.encode(channel).get, dataSuffixBits)
+
+  def decode[R](bv: ByteVector)(implicit codecR: Codec[R]): Seq[R] =
+    Codec.decode[Seq[R]](bv.bits).get.value
+
+  def transform(actions: List[HotStoreAction]): List[JournalResultTuple] =
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala]: I'm strongly opposed to this approach. the measuring mechanism should not be part of the main execution path. currently this code has to compute all the data (which is extremely expensive in some cases) even if the debug is not enabled.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+import scodec.Codec
+import scodec.bits.{BitVector, ByteVector}
+import coop.rchain.rspace.nextgenrspace.history.HistoryRepositoryImpl._
+
+object HistoryTransformers {
+
+  // TODO more readable if case class
+  type Result = (Blake2b256Hash, Option[PersistedData], HistoryAction)
+
+  final case class MeasureJournal(
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala]: could we spend some more time on this design?
key, type, action type, value seems to be less wasteful in this context.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    actions.map {
+      case i: InsertData[C, A] =>
+        val key      = hashDataChannel(i.channel)
+        val data     = encodeData(i.data)
+        val dataLeaf = DataLeaf(data)
+        val dataHash = Blake2b256Hash.create(data)
+        (
+          MeasureJournal(
+            key = key.bytes.toHex,
+            size = data.toBitVector.size,
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/HistoryTransformers.scala]: this is a bug from the original design - bits needs to be divided by 8 or `toByteVector.size` needs to be used.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala [hunk 4]
```diff
@@ -111,6 +112,46 @@ final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
...
+      case d: DeleteContinuations[C] =>
+        val key                                                      = hashContinuationsChannels(d.channels).bytes
+        val (dataSize, dataLen, continuationsSize, continuationsLen) = (0, 0, 0, 0)
+        s"${key.toHex};${key.length};${classOf[DeleteContinuations[C]].getName};$dataSize;$dataLen;$continuationsSize;$continuationsLen;"
+      case d: DeleteJoins[C] =>
+        val key                                                      = hashJoinsChannel(d.channel).bytes
+        val (dataSize, dataLen, continuationsSize, continuationsLen) = (0, 0, 0, 0)
+        s"${key.toHex};${key.length};${classOf[DeleteContinuations[C]].getName};$dataSize;$dataLen;$continuationsSize;$continuationsLen;"
+    }
+
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: suggestion:
```
  private def computeMeasure(actions: List[HotStoreAction]): List[String] =
    actions.map {
      case i: InsertData[C, A] =>
        val key  = hashDataChannel(i.channel).bytes
        val data = encodeData(i.data)
        s"${key.toHex};insert-data;${data.length};${i.data.length}"
      case i: InsertContinuations[C, P, K] =>
        val key  = hashContinuationsChannels(i.channels).bytes
        val data = encodeContinuations(i.continuations)
        s"${key.toHex};insert-continuation;${data.length};${i.continuations.length}"
      case i: InsertJoins[C] =>
        val key  = hashJoinsChannel(i.channel).bytes
        val data = encodeJoins(i.joins)
        s"${key.toHex};insert-join;${data.length}"
      case d: DeleteData[C] =>
        val key = hashDataChannel(d.channel).bytes
        s"${key.toHex};delete-data;0"
      case d: DeleteContinuations[C] =>
        val key = hashContinuationsChannels(d.channels).bytes
        s"${key.toHex};delete-continuation;0"
      case d: DeleteJoins[C] =>
        val key = hashJoinsChannel(d.channel).bytes
        s"${key.toHex};delete-join;0"
    }
```

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala [hunk 5]
```diff
@@ -192,6 +234,15 @@ object HistoryRepositoryImpl {
       .get
       .toByteVector
 
+  private def encodeUnsorted[D](data: Seq[D])(implicit codec: Codec[D]): ByteVector =
```
Comment (dzajkowski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: this looks out of place, shouldn't it be part of dev?

File: .gitignore [hunk 6]
```diff
@@ -70,6 +70,10 @@ node-client/.virtualenv
 \#*
 .ensime*
 ensime.sbt
+**/**/.bloop
```
Comment (dzajkowski) [inline .gitignore]: I know this isn't settled by a coding standard, but if we start adding every imaginable tool to the project ignore file - it's gonna get huge.
maybe instead 
https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore

Comment (kayvank) [reply .gitignore]: agreed. will modify

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala (hunk 5 above)
Comment (kayvank) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: yes. that is part of dev branch. strange!

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala (hunk 4 above)
Comment (kayvank) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: thank you.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala [hunk 7]
```diff
@@ -111,6 +112,38 @@ final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
 
   type Result = (Blake2b256Hash, Option[PersistedData], HistoryAction)
 
+  protected[this] val dataLogger: Logger = Logger("coop.rchain.rspace.datametrics")
+
+  private def measure(actions: List[HotStoreAction]): Unit =
```
Comment (lukasz-golebiewski) [inline rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: This is a side-effecting function so it should be enclosed in the context `F` and it's execution delayed

Comment (kayvank) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: wrapped the logger in context F as suggested.
rchain.coop.shared.Log has a very nice logger implementation that we can use but it requires adding the 'share' project as a dependency.

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala (hunk 5 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: `git rebase`?

**Comment by dzajkowski** [comment]
looks like something failed:
https://drone.rchain-dev.tk/rchain/rchain/3514/5

could you `bors try`?

**Comment by kayvank** [comment]
@dzajkowski a casper test failed as a result of JVM OOM. I reported to the #continuous-integration channel and @woky is working on it

**Comment by kayvank** [comment]
bors try

File: rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala (hunk 7 above)
Comment (lukasz-golebiewski) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: Isn't `shared` already a dependency of `rspace`? If not explicit then possibly transitive.
This sounds like something worth tackling in a separate PR

Comment (kayvank) [reply rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala]: yes

**Pull request merged**